File: attestation_asps/python_asps/copland.py
class ASP_PARAMS:
    def __init__(self, asp_id, asp_args, plc, targ_id):
        self.ASP_ID = asp_id
        self.ASP_ARGS = asp_args
        self.ASP_PLC = asp
        self.ASP_TARG_ID = asp_targ_id

class Attestation_Session:
    def __init__(self, session_plc, plc_mapping, pubkey_mapping):
        self.Session_Plc = session_plc
        self.Plc_Mapping = plc_mapping
        self.PubKey_Mapping = pubkey_mapping

class ProtocolRunRequest:
    def __init__(self, type, action, req_plc, term, rawev, attestation_session):
        self.TYPE = type
        self.ACTION = action
        self.REQ_PLC = req_plc
        self.TERM = term
        self.RAWEV = rawev
        self.ATTESTATION_SESSION = attestation_session

class ProtocolRunResponse:
    def __init__(self, type, action, success, payload):
        self.TYPE = type
        self.ACTION = action
        self.SUCCESS = success,
        self.PAYLOAD = payload

class ProtocolAppraiseRequest:
    def __init__(self, type, action, attestation_session, term, req_plc, evidence, rawev):
        self.TYPE = type,
        self.ACTION = action,
        self.ATTESTATION_SESSION = attestation_session,
        self.TERM = term,
        self.REQ_PLC = req_plc,
        self.EVIDENCE = evidence,
        self.RAWEV:  rawev

class ProtocolAppraiseResponse:
    def __init__(self, type, action, success, payload):
        self.TYPE = type
        self.ACTION = action
        self.SUCCESS = success
        self.PAYLOAD = payload

class RAWEV:
    def __init__(self, rawev):
        self.RawEv = rawev

    @staticmethod
    def from_json(dct):
        keys = dct.keys()
        if len(keys) == 1 and 'RawEv' in keys:
            return RAWEV(dct['RawEv'])
        else:
            return dct

class ASPRunRequest:
    def __init__(self, type, action, asp_id, asp_args, asp_plc, asp_targ_id, rawev):
        self.TYPE = type
        self.ACTION = action
        self.ASP_ID = asp_id
        self.ASP_ARGS = asp_args
        self.ASP_PLC = asp_plc
        self.ASP_TARG_ID = asp_targ_id
        self.RAWEV = rawev

    @staticmethod
    def from_json(dct):
        keys = dct.keys()
        if 'TYPE' in keys and 'ACTION' in keys and 'ASP_ID' in keys and 'ASP_ARGS' in keys and 'ASP_PLC' in keys and 'ASP_TARG_ID' in keys and 'RAWEV' in keys:
            return ASPRunRequest(dct['TYPE'], dct['ACTION'], dct['ASP_ID'], dct['ASP_ARGS'], dct['ASP_PLC'], dct['ASP_TARG_ID'], dct['RAWEV'])
        elif len(keys) == 1 and 'RawEv' in keys:
            return RAWEV.from_json(dct)
        else:
            return dct

class ASPRunResponse:
    def __init__(self, type, action, success, payload):
        self.TYPE = type
        self.ACTION = action
        self.SUCCESS = success
        self.PAYLOAD = payload

    @staticmethod
    def from_json(dct):
        keys = dct.keys()
        if 'TYPE' in keys and 'ACTION' in keys and 'SUCCESS' in keys and 'PAYLOAD' in keys:
            return ASPRunResponse(dct['TYPE'], dct['ACTION'], dct['SUCCESS'], dct['PAYLOAD'])
        elif len(keys) == 1 and 'RawEv' in keys:
            return RAWEV.from_json(dct)
        else:
            return dct


def failureASPRunResponse (error_msg):
    empty_evidence = RAWEV([])
    response = ASPRunResponse("RESPONSE",
                              "ASP_RUN",
                              False,
                              empty_evidence)
    return response

def successfulASPRunResponse (evidence):
    response = ASPRunResponse("RESPONSE",
                              "ASP_RUN",
                              True,
                              evidence)
    return response

File: lib/src/copland.rs
#![allow(non_camel_case_types)]
#![allow(non_snake_case)]

use anyhow::Result;
use base64::Engine;
use core::panic;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

type Plc = String;
type N_ID = String;
type ASP_ID = String;
type TARG_ID = String;
pub type ASP_ARGS = HashMap<String, String>;

#[derive(Serialize, Deserialize, Debug, Clone)]
struct ASP_PARAMS {
    ASP_ID: ASP_ID,
    ASP_ARGS: ASP_ARGS,
    ASP_PLC: Plc,
    ASP_TARG_ID: TARG_ID,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(tag = "FWD_CONSTRUCTOR", content = "FWD_BODY")]
enum FWD {
    COMP,
    ENCR,
    EXTD(String),
    KILL,
    KEEP,
}

#[derive(Serialize, Deserialize, Debug)]
#[serde(tag = "EVIDENCE_CONSTRUCTOR", content = "EVIDENCE_BODY")]
enum Evidence {
    mt,
    nn(N_ID),
    uu(Plc, FWD, ASP_PARAMS, Box<Evidence>),
    ss(Box<Evidence>, Box<Evidence>),
}

#[derive(Serialize, Deserialize, Debug, Clone)]
enum SP {
    ALL,
    NONE,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(tag = "ASP_CONSTRUCTOR", content = "ASP_BODY")]
enum ASP {
    NULL,
    CPY,
    ASPC(SP, FWD, ASP_PARAMS),
    SIG,
    HSH,
    ENC(Plc),
}

type Split = (SP, SP);

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(tag = "TERM_CONSTRUCTOR", content = "TERM_BODY")]
enum Term {
    asp(ASP),
    att(Plc, Box<Term>),
    lseq(Box<Term>, Box<Term>),
    bseq(Split, Box<Term>, Box<Term>),
    bpar(Split, Box<Term>, Box<Term>),
}

//type BS = bytestring::ByteString;

type RawEvT = Vec<String>; //Vec<BS>;

pub type EvidenceT = Vec<Vec<u8>>;

#[derive(Serialize, Deserialize, Debug)]
//#[serde(untagged)]
//#[serde(tag = "RawEv_CONSTRUCTOR", content = "RawEv_BODY")]
enum RawEv {
    RawEv(RawEvT),
}

#[derive(Serialize, Deserialize, Debug)]
struct ASPRunRequest {
    TYPE: String,
    ACTION: String,
    ASP_ID: String,
    ASP_ARGS: ASP_ARGS,
    ASP_PLC: Plc,
    ASP_TARG_ID: TARG_ID,
    RAWEV: RawEv,
}

#[derive(Serialize, Deserialize, Debug)]
struct ASPRunResponse {
    TYPE: String,
    ACTION: String,
    SUCCESS: bool,
    PAYLOAD: RawEv,
}

static APPRAISAL_SUCCESS_RESPONSE: &str = "";

fn successfulASPRunResponse(evidence: RawEv) -> ASPRunResponse {
    ASPRunResponse {
        TYPE: "RESPONSE".to_string(),
        ACTION: "ASP_RUN".to_string(),
        SUCCESS: true,
        PAYLOAD: evidence,
    }
}

// Currently the reason string is ignored, but eventually
// should be incorporated into the response.
fn failureASPRunResponse(_reason: String) -> ASPRunResponse {
    eprintln!("Error: {_reason}");

    ASPRunResponse {
        TYPE: "RESPONSE".to_string(),
        ACTION: "ASP_RUN".to_string(),
        SUCCESS: false,
        PAYLOAD: RawEv::RawEv(Vec::new()),
    }
}

// NOTE: This function will exit the process with a status code of 1
fn respond_with_failure(reason: String) -> ! {
    let resp_json = serde_json::to_string(&failureASPRunResponse(reason)).unwrap_or_else(|error| {
        panic!("Failed to json.encode failure response: {error:?}");
    });
    println!("{resp_json}");
    std::process::exit(1);
}

// Convert base64 encoded string to vec u8
fn base64_to_vec(base64: &str) -> Vec<u8> {
    let bytes = base64::engine::general_purpose::STANDARD
        .decode(base64)
        .unwrap_or_else(|error| {
            respond_with_failure(format!("Failed to decode base64: {error:?}"));
        });
    bytes
}

fn vec_to_base64(vec: &[u8]) -> String {
    base64::engine::general_purpose::STANDARD.encode(vec)
}

fn rawev_to_vec(rawev: RawEv) -> Vec<Vec<u8>> {
    match rawev {
        RawEv::RawEv(rawevt) => rawevt.iter().map(|base64| base64_to_vec(&base64)).collect(),
    }
}

fn vec_to_rawev(vec: Vec<Vec<u8>>) -> RawEv {
    RawEv::RawEv(vec.iter().map(|bytes| vec_to_base64(bytes)).collect())
}

fn gather_args_and_req() -> (EvidenceT, ASP_ARGS) {
    let args: Vec<String> = std::env::args().collect();

    if args.len() != 2 {
        eprintln!("Usage: {} <ASPRunRequest JSON>", args[0]);
        respond_with_failure("Invalid arguments to ASP".to_string());
    }

    let json_req = &args[1];
    let req: ASPRunRequest = serde_json::from_str(json_req).unwrap_or_else(|error| {
        respond_with_failure(format!("Failed to parse ASPRunRequest: {error:?}"));
    });

    (rawev_to_vec(req.RAWEV), req.ASP_ARGS)
}

pub fn handle_appraisal_body(body: fn(EvidenceT, ASP_ARGS) -> Result<Result<()>>) -> ! {
    let (ev, args) = gather_args_and_req();
    match body(ev, args) {
        Ok(appr_res) => match appr_res {
            Ok(_) => {
                let response =
                    successfulASPRunResponse(RawEv::RawEv(vec![APPRAISAL_SUCCESS_RESPONSE.into()]));
                let resp_json = serde_json::to_string(&response).unwrap_or_else(|error| {
                    respond_with_failure(format!("Failed to json.encode response: {error:?}"));
                });
                println!("{resp_json}");
                std::process::exit(0);
            }
            Err(reason) => {
                // This is not a FAILURE, but rather an APPRAISAL that ended in a negative result.
                let response = successfulASPRunResponse(RawEv::RawEv(vec![reason.to_string()]));
                let resp_json = serde_json::to_string(&response).unwrap_or_else(|error| {
                    respond_with_failure(format!("Failed to json.encode response: {error:?}"));
                });
                println!("{resp_json}");
                std::process::exit(0);
            }
        },
        Err(reason) => {
            respond_with_failure(reason.to_string());
        }
    }
}

pub fn handle_body(body: fn(EvidenceT, ASP_ARGS) -> Result<EvidenceT>) -> ! {
    let (ev, args) = gather_args_and_req();
    match body(ev, args) {
        Ok(ev) => {
            let response = successfulASPRunResponse(vec_to_rawev(ev));
            let resp_json = serde_json::to_string(&response).unwrap_or_else(|error| {
                respond_with_failure(format!("Failed to json.encode response: {error:?}"));
            });
            println!("{resp_json}");
            std::process::exit(0);
        }
        Err(reason) => {
            respond_with_failure(reason.to_string());
        }
    }
}
